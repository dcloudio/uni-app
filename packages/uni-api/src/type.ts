type Data = Record<string, unknown>
type ProtocolConstructor<T = any> =
  | { new (...args: any[]): T & object }
  | { (): T }
  | ProtocolMethod<T>

type ProtocolMethod<T, TConstructor = any> = T extends (...args: any) => any
  ? { new (): TConstructor; (): T; readonly prototype: TConstructor }
  : never
type ProtocolType<T> = ProtocolConstructor<T> | ProtocolConstructor<T>[]

type ApiArgsValidator<T, P> = (value: T, params: P) => boolean | string | void

type ApiProtocol<T extends ApiLike, P = AsyncApiOptions<T>> = {
  [K in keyof P]: ProtocolType<P[K]> | ProtocolOptions<P[K]>
}

type ApiProtocols<T extends ApiLike> = ApiProtocol<T> | ProtocolOptions[]

interface ApiOptions<T extends ApiLike, P = AsyncApiOptions<T>> {
  beforeInvoke?: (args: unknown) => boolean | void | string
  beforeAll?: (res: unknown) => void
  beforeSuccess?: (res: unknown, args: P) => void
  formatArgs?: {
    [K in keyof P]?: ApiArgsValidator<P[K], P> | P[K]
  }
}

interface ProtocolOptions<T = any> {
  name?: string
  type?: ProtocolType<T>
  required?: boolean
  validator?(value: T): boolean | void | string
}

interface AsyncMethodOptionLike {
  success?: (...args: any[]) => void
}

type PromisifySuccessResult<P, R> = P extends {
  success: any
}
  ? void
  : P extends { fail: any }
  ? void
  : P extends { complete: any }
  ? void
  : Promise<R>
type ApiLike = (...args: any[]) => any

type TaskApiLike = ApiLike

type AsyncApiLike = (args: any) => Promise<unknown> | void

type AsyncApiOptions<T extends ApiLike> = Required<Parameters<T>>[0]

type AsyncApiRes<T extends AsyncMethodOptionLike> = Omit<
  Parameters<Exclude<T['success'], undefined>>[0],
  'errMsg' | 'errSubject'
>

type AsyncApiRequired<T extends AsyncMethodOptionLike> = <P extends T>(
  args: P
) => PromisifySuccessResult<P, AsyncApiRes<T>>

type AsyncApiOptional<T extends AsyncMethodOptionLike> = <P extends T>(
  args?: P
) => PromisifySuccessResult<P, AsyncApiRes<T>>

interface AsyncApiOptionalOptions {
  success?: any
  fail?: any
  complete?: any
}

type AsyncApi<T extends AsyncMethodOptionLike> =
  AsyncApiOptionalOptions extends T ? AsyncApiOptional<T> : AsyncApiRequired<T>
