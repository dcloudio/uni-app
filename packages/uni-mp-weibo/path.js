const path = require('path')
module.exports = {
  src: path.join(__dirname, 'src'),
  lib: path.join(__dirname, 'lib')
}
