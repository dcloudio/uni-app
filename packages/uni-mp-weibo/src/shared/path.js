export function parsePath (path) {
  let hash = ''
  let query = ''

  const hashIndex = path.indexOf('#')
  if (hashIndex >= 0) {
    hash = path.slice(hashIndex)
    path = path.slice(0, hashIndex)
  }

  const queryIndex = path.indexOf('?')
  if (queryIndex >= 0) {
    query = path.slice(queryIndex + 1)
    path = path.slice(0, queryIndex)
  }

  return {
    path,
    query,
    hash
  }
}
