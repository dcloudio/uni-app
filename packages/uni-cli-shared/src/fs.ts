import fs from 'fs'
import path from 'path'

export function emptyDir(dir: string, skip: string[] = []) {
  for (const file of fs.readdirSync(dir)) {
    if (skip.includes(file)) {
      continue
    }
    // node >= 14.14.0
    fs.rmSync(path.resolve(dir, file), { recursive: true, force: true })
  }
}
