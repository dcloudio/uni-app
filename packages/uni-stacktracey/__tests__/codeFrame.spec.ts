import fs from 'fs'
import path from 'path'

const {
  SourceMapConsumer,
  generateCodeFrameSourceMapConsumer,
  generateCodeFrameWithSourceMapPath,
} = require('../dist/uni-stacktracey.cjs.js')
import type {
  BasicSourceMapConsumer,
  IndexedSourceMapConsumer,
} from '../lib/source-map'

const warnings = [
  {
    type: 'warning',
    message: "'DATA: String' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 27,
    column: 71,
  },
  {
    type: 'warning',
    message:
      'Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions',
    file: 'index.kt',
    line: 41,
    column: 25,
  },
  {
    type: 'warning',
    message: "Condition 'cursor == null' is always 'false'",
    file: 'index.kt',
    line: 61,
    column: 13,
  },
  {
    type: 'warning',
    message: "'DATA: String' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 63,
    column: 78,
  },
  {
    type: 'warning',
    message: "Variable 'width' initializer is redundant",
    file: 'index.kt',
    line: 69,
    column: 21,
  },
  {
    type: 'warning',
    message: "Variable 'height' initializer is redundant",
    file: 'index.kt',
    line: 70,
    column: 22,
  },
  {
    type: 'warning',
    message:
      'Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions',
    file: 'index.kt',
    line: 97,
    column: 47,
  },
  {
    type: 'warning',
    message:
      "'getter for defaultDisplay: Display!' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 122,
    column: 44,
  },
  {
    type: 'warning',
    message: "'getRealSize(Point!): Unit' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 123,
    column: 24,
  },
  {
    type: 'warning',
    message:
      "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 152,
    column: 51,
  },
  {
    type: 'warning',
    message:
      "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
    file: 'index.kt',
    line: 153,
    column: 47,
  },
]

const filename = path.resolve(__dirname, 'index.kt.map')

describe('code-frame', () => {
  let consumer: BasicSourceMapConsumer | IndexedSourceMapConsumer
  async function initConsumer() {
    if (!consumer) {
      consumer = await new SourceMapConsumer(fs.readFileSync(filename, 'utf8'))
    }
    return consumer
  }

  test('generateCodeFrame', async () => {
    const consumer = await initConsumer()
    warnings.forEach((w) => {
      expect(
        generateCodeFrameSourceMapConsumer(consumer, w, {
          sourceRoot: '/Users/fxy/Projects/Gitcode/hello-uts',
        })
      ).toMatchSnapshot()
    })
  })

  test('generateCodeFrameWithSourceMapPath', async () => {
    ;(
      await generateCodeFrameWithSourceMapPath(
        filename,
        JSON.stringify(warnings),
        { sourceRoot: '/Users/fxy/Projects/Gitcode/hello-uts' }
      )
    ).forEach((m: unknown) => {
      expect(m).toMatchSnapshot()
    })
  })
})
