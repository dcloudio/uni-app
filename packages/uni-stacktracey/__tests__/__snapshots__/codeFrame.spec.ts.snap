// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code-frame generateCodeFrame 1`] = `
Object {
  "code": "36 |   */
37 |  const MEDIA_PROJECTIONS_API_16 = arrayOf(
38 |    MediaStore.Images.ImageColumns.DATA,
   |                                   ^
39 |    MediaStore.Images.ImageColumns.DATE_TAKEN,
40 |    MediaStore.Images.ImageColumns.WIDTH,",
  "column": 33,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 38,
  "message": "'DATA: String' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 2`] = `
Object {
  "code": "76 |        }
77 |        lastFileObserverTime = System.currentTimeMillis()
78 |        listenOption!.onImageCatchChange(newPath)
   |        ^
79 |      }
80 |    }",
  "column": 6,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 78,
  "message": "Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 3`] = `
Object {
  "code": "124|  
125|  
126|      if (cursor == null) {
   |          ^
127|        return
128|      }",
  "column": 8,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 126,
  "message": "Condition 'cursor == null' is always 'false'",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 4`] = `
Object {
  "code": "133|  
134|      // 获取各列的索引
135|      let dataIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
   |                                                                           ^
136|      let dateTakenIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATE_TAKEN)
137|      let widthIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.WIDTH)",
  "column": 73,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 135,
  "message": "'DATA: String' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 5`] = `
Object {
  "code": "141|      let data = cursor.getString(dataIndex)
142|      let dateTaken = cursor.getLong(dateTakenIndex)
143|      let width = 0;
   |                   ^
144|      let height = 0;
145|      if (widthIndex >= 0 && heightIndex >= 0) {",
  "column": 17,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 143,
  "message": "Variable 'width' initializer is redundant",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 6`] = `
Object {
  "code": "142|      let dateTaken = cursor.getLong(dateTakenIndex)
143|      let width = 0;
144|      let height = 0;
   |                    ^
145|      if (widthIndex >= 0 && heightIndex >= 0) {
146|        width = cursor.getInt(widthIndex)",
  "column": 18,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 144,
  "message": "Variable 'height' initializer is redundant",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 7`] = `
Object {
  "code": "180|    if (checkScreenShot(data, dateTaken, width, height)) {
181|      if (!checkCallback(data)) {
182|        listenOption!.onImageCatchChange(data)
   |        ^
183|      }
184|    } else {",
  "column": 6,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 182,
  "message": "Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 8`] = `
Object {
  "code": "248|    try {
249|      let windowManager = getUniActivity()!.getSystemService(Context.WINDOW_SERVICE) as WindowManager
250|      let defaultDisplay = windowManager.defaultDisplay
   |                                         ^
251|      defaultDisplay.getRealSize(screenSize)
252|    } catch (e) {",
  "column": 39,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 250,
  "message": "'getter for defaultDisplay: Display!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 9`] = `
Object {
  "code": "249|      let windowManager = getUniActivity()!.getSystemService(Context.WINDOW_SERVICE) as WindowManager
250|      let defaultDisplay = windowManager.defaultDisplay
251|      defaultDisplay.getRealSize(screenSize)
   |                     ^
252|    } catch (e) {
253|      e.printStackTrace()",
  "column": 19,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 251,
  "message": "'getRealSize(Point!): Unit' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 10`] = `
Object {
  "code": "308|      var directory_screenshot: File;
309|  
310|      var directory_pictures = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_PICTURES);
   |                                                ^
311|      var directory_dcim = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_DCIM);
312|  ",
  "column": 46,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 310,
  "message": "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrame 11`] = `
Object {
  "code": "309|  
310|      var directory_pictures = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_PICTURES);
311|      var directory_dcim = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_DCIM);
   |                                            ^
312|  
313|      if (Build.MANUFACTURER.equals(\\"Xiaomi\\", true)) {",
  "column": 42,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 311,
  "message": "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 1`] = `
Object {
  "code": "36 |   */
37 |  const MEDIA_PROJECTIONS_API_16 = arrayOf(
38 |    MediaStore.Images.ImageColumns.DATA,
   |                                   ^
39 |    MediaStore.Images.ImageColumns.DATE_TAKEN,
40 |    MediaStore.Images.ImageColumns.WIDTH,",
  "column": 33,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 38,
  "message": "'DATA: String' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 2`] = `
Object {
  "code": "76 |        }
77 |        lastFileObserverTime = System.currentTimeMillis()
78 |        listenOption!.onImageCatchChange(newPath)
   |        ^
79 |      }
80 |    }",
  "column": 6,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 78,
  "message": "Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 3`] = `
Object {
  "code": "124|  
125|  
126|      if (cursor == null) {
   |          ^
127|        return
128|      }",
  "column": 8,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 126,
  "message": "Condition 'cursor == null' is always 'false'",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 4`] = `
Object {
  "code": "133|  
134|      // 获取各列的索引
135|      let dataIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
   |                                                                           ^
136|      let dateTakenIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATE_TAKEN)
137|      let widthIndex = cursor.getColumnIndex(MediaStore.Images.ImageColumns.WIDTH)",
  "column": 73,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 135,
  "message": "'DATA: String' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 5`] = `
Object {
  "code": "141|      let data = cursor.getString(dataIndex)
142|      let dateTaken = cursor.getLong(dateTakenIndex)
143|      let width = 0;
   |                   ^
144|      let height = 0;
145|      if (widthIndex >= 0 && heightIndex >= 0) {",
  "column": 17,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 143,
  "message": "Variable 'width' initializer is redundant",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 6`] = `
Object {
  "code": "142|      let dateTaken = cursor.getLong(dateTakenIndex)
143|      let width = 0;
144|      let height = 0;
   |                    ^
145|      if (widthIndex >= 0 && heightIndex >= 0) {
146|        width = cursor.getInt(widthIndex)",
  "column": 18,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 144,
  "message": "Variable 'height' initializer is redundant",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 7`] = `
Object {
  "code": "180|    if (checkScreenShot(data, dateTaken, width, height)) {
181|      if (!checkCallback(data)) {
182|        listenOption!.onImageCatchChange(data)
   |        ^
183|      }
184|    } else {",
  "column": 6,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 182,
  "message": "Unnecessary non-null assertion (!!) on a non-null receiver of type onImageCatchOptions",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 8`] = `
Object {
  "code": "248|    try {
249|      let windowManager = getUniActivity()!.getSystemService(Context.WINDOW_SERVICE) as WindowManager
250|      let defaultDisplay = windowManager.defaultDisplay
   |                                         ^
251|      defaultDisplay.getRealSize(screenSize)
252|    } catch (e) {",
  "column": 39,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 250,
  "message": "'getter for defaultDisplay: Display!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 9`] = `
Object {
  "code": "249|      let windowManager = getUniActivity()!.getSystemService(Context.WINDOW_SERVICE) as WindowManager
250|      let defaultDisplay = windowManager.defaultDisplay
251|      defaultDisplay.getRealSize(screenSize)
   |                     ^
252|    } catch (e) {
253|      e.printStackTrace()",
  "column": 19,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 251,
  "message": "'getRealSize(Point!): Unit' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 10`] = `
Object {
  "code": "308|      var directory_screenshot: File;
309|  
310|      var directory_pictures = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_PICTURES);
   |                                                ^
311|      var directory_dcim = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_DCIM);
312|  ",
  "column": 46,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 310,
  "message": "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;

exports[`code-frame generateCodeFrameWithSourceMapPath 11`] = `
Object {
  "code": "309|  
310|      var directory_pictures = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_PICTURES);
311|      var directory_dcim = File(Environment.getExternalStorageDirectory(), Environment.DIRECTORY_DCIM);
   |                                            ^
312|  
313|      if (Build.MANUFACTURER.equals(\\"Xiaomi\\", true)) {",
  "column": 42,
  "file": "uni_modules/uts-screenshot-listener/utssdk/app-android/index.uts",
  "line": 311,
  "message": "'getExternalStorageDirectory(): File!' is deprecated. Deprecated in Java",
  "type": "warning",
}
`;
